prov/util/src/util_mem_hooks.c

```diff
@@ -373,6 +381,85 @@ static bool ofi_is_function_patched(struct ofi_intercept *intercept)
+ *
+ * @param[in] _reg  register number (0-31), @param[out] _reg register number (0-31), @param[imm] 12 bit immmediate value
+ */
+#define JALR(_regs, _regd, _imm) (((_imm) << 20) | ((_regs) << 15) | (0b000 << 12) | ((_regd) << 7) | (0x67))
+
+/**
+ * @brief ADDI - Add 12 bit immediate to source register, save to destination register 
+ *
+ * @param[in] _reg  register number (0-31), @param[out] _reg register number (0-31), @param[imm] 12 bit immmediate value
+ */
+#define ADDI(_regs, _regd, _imm) (((_imm) << 20) | ((_regs) << 15) | (0b000 << 12) | ((_regd) << 7) | (0x13))
+#define ADD(_regs_a, _regs_b, _regd) ((_regs_b << 20) | (_regs_a << 15) | (0b000 << 12) | ((_regd) << 7) | (0x33))
+
+/**
+ * @brief LUI - load upper 20 bit immediate to destination register
+ *
+ * @param[in] _reg  register number (0-31), @param[out] _reg register number (0-31), @param[imm] 12 bit immmediate value
+ */
+#define LUI(_regd, _imm) (((_imm) << 12) | ((_regd) << 7) | (0x37))
+
+/**
+ * @brief SLLI - left-shift immediate number of bits in source register into destination register
+ *
+ * @param[in] _reg  register number (0-31), @param[out] _reg register number (0-31), @param[imm] 12 bit immmediate value
+ */
+#define SLLI(_regs, _regd, _imm) (((_imm) << 20) | ((_regs) << 15) | (0b001 << 12) | ((_regd) << 7) | (0x13))
+
+static int ofi_patch_function(struct ofi_intercept *intercept)
+{
+    /*
+     * r15 is the highest numbered temporary register. I am
+     * assuming this one is safe to use.
+     */
+    uintptr_t addr __attribute__ ((aligned(32))) = (uintptr_t) intercept->patch_data;
+    uintptr_t value __attribute__ ((aligned(32))) = (uintptr_t) intercept->our_func;
+
+    *(uint32_t*) (addr +  0) = LUI  (X31, ((0xFFFFF << 12) & ( ((value) >> 32) + 1 ) ) >> 12);
+    *(uint32_t*) (addr +  4) = ADDI (X31, X31, ((0xFFF)    & ( ((value) >> 32) + 1 ) )      );
+    *(uint32_t*) (addr +  8) = LUI  (X30, ((0xFFFFF << 12) & ( (((value)) + 1)     ) ) >> 12);
+    *(uint32_t*) (addr + 12) = SLLI (X31, X31, 32);
+    *(uint32_t*) (addr + 16) = ADD  (X30, X31, X31);
+    *(uint32_t*) (addr + 20) = JALR (X31, X0, ((0xFFF)     & ( ((value)) + 1)));
+
+    intercept->patch_data_size = 24;
+
+    return ofi_apply_patch(intercept);
+}
+
+/*
+ * Please see comments at other ofi_is_function_patched() function
+ */
+static bool ofi_is_function_patched(struct ofi_intercept *intercept)
+{
+    uintptr_t addr = (uintptr_t) intercept->orig_func;
+    /*
+     * focus on the instructions in the bytes.
+     */
+    return (
+        ((*(uint32_t *) (addr +  0)) & 0xFF) == 0x37 &&
+        ((*(uint32_t *) (addr +  4)) & 0xFF) == 0x13 &&
+        ((*(uint32_t *) (addr +  8)) & 0xFF) == 0x37 &&
+        ((*(uint32_t *) (addr + 12)) & 0xFF) == 0x13 &&
+        ((*(uint32_t *) (addr + 16)) & 0xFF) == 0x33 &&
+        ((*(uint32_t *) (addr + 20)) & 0xFF) == 0x67
+	);
+}
```

Comment:
Did the above code come from OpenMPI?  If so, can you capture that information in the commit message?

Reply:
That code did not come from MPI, it's from another network library I've done work on; I'll update that information in the source and in a commit message.

Reply:
If the source was from another library, we need to verify that the license is appropriate for use.

Reply:
understood, I'll put together a different implementation.